### ISSUE TITLE
No need to use transferFrom to solve ex18

### ISSUE THREAD

**Issue by Baraa42** [l-henri/starknet-erc20#11]
https://github.com/l-henri/starknet-erc20/blob/4f8fa26c773ff0c7c2f73a6b2f7b544a374677f7/contracts/Evaluator.cairo#L595-

The exercise ask to use transferFrom, there is no need to do it if the token is the _contractToken_, you just burn the amount immediately, also I don't think the approve is needed. Here is the code I used to solve the exercise .

```
@external
func withdraw_all_tokens{
        syscall_ptr : felt*, 
        pedersen_ptr : HashBuiltin*,
        range_check_ptr
    }() -> (amount : Uint256):
    alloc_locals
    # get caller address
    let (caller) = get_caller_address()
    # get amount in custody
    let (amount) = tokens_in_custody(caller)
    # get DTK address
    let (read_dtk_address) = dummy_token.read()
    # approve caller and transfer DTK
    IDTKERC20.approve(read_dtk_address,caller, amount)
    IDTKERC20.transfer(read_dtk_address,caller, amount)
    # Update mapping
    let zero : Uint256 = Uint256(0, 0)
    tokens_in_custody_mapping.write(caller, zero)
    # burn tokens
    ERC20_burn(caller, amount)
    return(amount)
end
```

_Note: I could have used `amount = balanceOf(DTK, caller)` instead but left the code I used for previous exercises._

**Comment by trevis-dev** [comment]
Hello ! 

That may be true, but it is not the objective of the exercise. So, while is not enforced in the exercise's logic, can you do it with another Token contract ?

Then, as it could still technically be done without "approve", I suppose users wouldn't use these contracts if they didn't have to approve tranfers first, right?

**Comment by Baraa42** [comment]
Yes you are correct, but it's easier to do it without transferFrom, I would suggest adding a check on `submitted_exercise_token_address` to be different than `sender_address` so that people will have to conform to exercise direction.
